File: src/sub/rmsdatvar.c
#include <math.h>
#include <stdbool.h>
#include <stdio.h>

#define IEQ 658
#define IST 650

extern int legs;
extern int nitt;
extern int isingle;
extern int nobswithw0;
extern int nvareff;
extern bool single_turbo;
extern int knobs[IEQ];
extern int iyr[IEQ];
extern int imo[IEQ];
extern int iday[IEQ];
extern int ihr[IEQ];
extern int imin[IEQ];
extern double w[IST][IEQ];
extern double res[IST][IEQ];
extern double avres[IEQ];
extern double rms[IEQ];
extern double davar1;
extern double xmsqrs1;
extern FILE *fm_ptr;

void rmsdatvar(void) {
    int knobst = 0;
    double tres = 0.0f;
    FILE *logfp = (fm_ptr != NULL) ? fm_ptr : stdout;

    davar1 = 0.0f;
    xmsqrs1 = 0.0f;

    for (int i = 0; i < legs; ++i) {
        int j2 = knobs[i];
        knobst += j2;

        for (int j = 0; j < j2; ++j) {
            if (w[j][i] <= 0.0f) {
                continue;
            }
            avres[i] += res[j][i] * w[j][i];
            rms[i] += (res[j][i] * res[j][i]) * w[j][i] * w[j][i];
        }

        davar1 += rms[i];
        tres += avres[i];

        if ((j2 - nobswithw0) <= 1) {
            continue;
        }

        rms[i] = sqrtf((rms[i] - (avres[i] * avres[i]) / (j2 - nobswithw0))
                     / ((j2 - nobswithw0) - 1));
        avres[i] = avres[i] / (j2 - nobswithw0);
    }

    if (nitt == 0 && !single_turbo) {
        fprintf(logfp, "\n Total number of observations is: %5d\n\n", knobst);
    }

    if (nitt == 0 && isingle != 0 && !single_turbo) {
        fprintf(logfp, "Number of observations with normalized weight 0.0 : %d\n", nobswithw0);
        fprintf(logfp, "knobs(i)   = %d\n", knobst);
        fprintf(logfp, "nobswithw0 = %d\n", nobswithw0);
    }

    if ((knobst - nobswithw0) > nvareff) {
        davar1 = (davar1 - tres * tres / (knobst - nobswithw0))
               / ((knobst - nobswithw0) - nvareff);
        xmsqrs1 = davar1 * ((knobst - nobswithw0) - nvareff) / (knobst - nobswithw0);
    } else {
        davar1 = 999.99f;
        xmsqrs1 = 999.99f;
    }

    if (!single_turbo && isingle == 0) {
        fprintf(logfp, "\n$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n");
        fprintf(logfp, "Events with  | AVRES | > 1.0 SEC are suspicious !\n\n");

        int found = 0;
        for (int i = 0; i < legs; ++i) {
            if (fabs(avres[i]) > 1.0f) {
                fprintf(logfp,
                        " Event# %3d >>> %02d%02d%02d %02d%02d  AVRES =%6.2f NOBS =%3d\n",
                        i + 1, iyr[i], imo[i], iday[i], ihr[i], imin[i], avres[i], knobs[i]);
                found++;
            }
        }

        if (found > 0) {
            fprintf(logfp, "\n^^^^^^^^^ C H E C K   these events above ^^^^^^^^\n");
            fprintf(logfp, "$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n");
        } else {
            fprintf(logfp, "ZERO events of this kind found! (lucky guy!!!)\n");
        }
        fprintf(logfp, "\n");
    }
}

File: src_fortran/sub/rmsdatvar.f
c
c
c
c
c
c
      subroutine RMSDATVAR
c
c     compute RMS and DATVAR for all the events (EQs & shots)
c
      implicit none
      include '../inc/vel_com.inc'
      integer jj,i,j2,j,knobst
      real tres
c
c
c   now calculate rms for all events
c
      jj=0
      davar1=0.0
      xmsqrs1=0.0
      tres=0.0
      knobst=0
      do 11 i=1,legs
         j2=knobs(i)
         knobst=knobst+j2
         do 12 j=1,j2
            if(w(j,i).le.0.0) goto 12
            avres(i)=avres(i)+res(j,i)*w(j,i)
Cek  changed by ek to res*w*res*w::::>
            rms(i)=rms(i)+ (res(j,i)*res(j,i))*w(j,i)*w(j,i)
 12      continue
         davar1=davar1+rms(i)
         tres=tres+avres(i)
cek next statement
         if( (j2-nobswithw0) .le.1) goto 11
         rms(i)=sqrt( (rms(i)-avres(i)**2/(j2-nobswithw0) )
     &               /( (j2-nobswithw0) -1))
         avres(i)=avres(i)/(j2-nobswithw0)
 11   continue
      if(nitt.eq.0)then
         if(.not.single_turbo)then
            write(16,310) knobst
         endif
      endif
310   format(/,' Total number of observations is: ',i5,/)
      if(nitt.eq.0.and.isingle.ne.0)then
         if(.not.single_turbo)then
            write(16,*)'Number of observations with '//
     &                 'normalized weight 0.0 :', nobswithw0
            write(16,*)'knobs(i)   = ',knobst
            write(16,*)'nobswithw0 = ',nobswithw0
         endif
      endif
c
c     calculate data variance:
c
c     NVAReff is the actual number of unknowns to solve for !!!
c
      if( (knobst-nobswithw0).gt.nvareff)then
        davar1=(davar1-tres*tres/(knobst-nobswithw0))
     &          / ((knobst-nobswithw0)-nvareff)
        xmsqrs1=davar1*((knobst-nobswithw0)-nvareff)/(knobst-nobswithw0)
      else
         davar1=999.99
         xmsqrs1=999.99
      endif
c
      i f (.not.single_turbo) t h e n
      if(isingle.eq.0)then
         write(16,*)
         write(16,*)'$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$'
         write(16,*)'Events with  | AVRES | > 1.0 SEC are suspicious !'
         write(16,*)
         j=0
         do i=1,legs
            if(abs(avres(i)).gt.1.0)then
               write(16,'('' Event# '',i3,'' >>> '',1x,3i2.2,1x,2i2.2,
     &                    ''  AVRES ='',f6.2,'' NOBS ='',i3)')
     &                    i,iyr(i),imo(i),iday(i),ihr(i),imin(i),
     &                    avres(i),knobs(i)
               j=j+1
            endif
         enddo
         if(j.gt.0)then
         write(16,*)
         write(16,*)'^^^^^^^^^ C H E C K   these events above ^^^^^^^^'
         write(16,*)'$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$'
         else
            write(16,*)'ZERO events of this kind found! (lucky guy!!!)'
         endif
         write(16,*)
      endif
      e n d i f
c
      return
      end ! of subr. rmsdatvar
